File: C++/프로그래머스/2/87946. 피로도/피로도.cpp
#include <vector>

using namespace std;

int answer;

void dfs(int k, int count, vector<bool> visited, vector<vector<int>> dungeons) {
    for(int i = 0; i < dungeons.size(); i++) {
        if(!visited[i] && k  >= dungeons[i][1] && dungeons[i][0] <= k) {
            visited[i] = true;
            dfs(k - dungeons[i][1], count + 1, visited, dungeons);
            visited[i] = false;
        }
    }
    if(answer < count) answer = count;
}

int solution(int k, vector<vector<int>> dungeons) {
    dfs(k, 0, vector<bool>(dungeons.size()), dungeons);
    return answer;
}

File: Python3/프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = 0
    
    def dfs(k, count, visited):
        for i, (n, u) in enumerate(dungeons):
            if not visited[i] and n <= k and k - u >= 0:
                visited[i] = True
                dfs(k - u, count + 1, visited)
                visited[i] = False
            nonlocal answer
            answer = max(answer, count)
    dfs(k, 0, [0]  * len(dungeons))
    return answer